### PR TITLE
post on integrating Salt and Vault

### PR DESCRIPTION
**Pull request by thraxil**

### CODE REVIEW

File: content/articles/using-vault-as-an-external-pillar-for-salt.md [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+           bar
+
+This is a working setup.
+Salt is accessing pillar data that is securely stored in Vault.
+From here out, I just want to make some clarifications and suggest some improvements.
+
+One thing that caught me by surprise was that Salt uses a single Vault entry to store all the data for the external pillar rather than making use of the hierarchical path structure that Vault provides.
+You can set up multiple external pillars by adding additional `- vault: vault_config path=secret/salt/pillar_data` lines to the config and changing the path.
+This is particularly useful if different groups of admins might be responsible for different areas and you want to be able to further segregate and limit access.
+Adding those requires changing the Salt config and restarting the Salt master each time, so you probably don't want to have to explictly configure different paths for every secret that Vault is storing.
```
Comment (nikolas) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: s/explictly/explicitly

File: content/articles/using-vault-as-an-external-pillar-for-salt.md [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+We have dozens of secrets stored in ours in a nested structure (Django server secrets, AWS credential pairs, database passwords, etc.).
+Editing the `pillar_data` entry involves writing the data out to a JSON file, extracting the `data` key from that, editing the contents of that key, saving that new JSON object to a file, then writing it back to vault:
+
+    $ vault read -format=json secret/salt/pillar_data > data.json
+    $ emacs data.json
+	# ... do a bunch of complicated edits to raw JSON ...
+	$ vault write secret/salt/pillar_data @data.json
+
+JSON is not exactly a friendly format for editing by hand.
+It's way too easy to mess it up and it can be very difficult to see the structurewhen there are more than a few entries.
```
Comment (nikolas) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: "structurewhen" -- need a space between these words

File: content/articles/using-vault-as-an-external-pillar-for-salt.md [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+[INFO    ] Loading fresh modules for state activity
+local:
+    Data failed to compile:
+----------
+    Pillar failed to render with the following messages:
+----------
+    Failed to load ext_pillar vault: [Errno 111] Connection refused
+```
+
+Interestingly, `saltutil.refresh_pillar` didn't explictly through an error, but it did result in an entry in the salt master log:
```
Comment (nikolas) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: s/through/throw

File: content/articles/using-vault-as-an-external-pillar-for-salt.md [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+[INFO    ] Loading fresh modules for state activity
+local:
+    Data failed to compile:
+----------
+    Pillar failed to render with the following messages:
+----------
+    Failed to load ext_pillar vault: HTTP 503: Service Unavailable
+```
+
+Unsealing Vault immediately fixed the issues without requring any other action.
```
Comment (nikolas) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: s/requring/requiring

File: content/articles/using-vault-as-an-external-pillar-for-salt.md (hunk 3 above)
Comment (nikolas) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: also: explictly -> explicitly

File: content/articles/using-vault-as-an-external-pillar-for-salt.md [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+We've been trying to up our security game lately.
+One of the areas we identified as needing improvement was storage of credentials, keys, and passwords that our configuration management uses.
+
+We use [Salt](https://saltstack.com/) as our configuration management tool of choice. Salt already lets you separate your setup into 'states' and the 'pillar' where states describe how various components are deployed and configured and the pillar is a relatively simple data structure of values that can be plugged into the states when they are run.
+At a very high level, you can think of Salt states as functions and the pillar data as the parameters that are passed to those functions.
+
+It's already a best practice to keep the states and pillar data in separate git repos and to keep sensitive data out of the state definitions.
+Even if the pillar repo isn't any more secure than the states repo, that simplifies basic tasks like changing passwords or rotating credentials.
+
+But, we still end up with the pillar repo containing more sensitive information unencrypted in one location than we were really comfortable with.
```
Comment (sdreher) [inline content/articles/using-vault-as-an-external-pillar-for-salt.md]: tense match: "ended up with"

**Pull request merged**